[core/string] Add String#unicode_normalize specs

JuanitoFatas opened the pull request:
This pull requests adds [`String#unicode_normalize`](http://ruby-doc.org/stdlib-2.2.0/libdoc/unicode_normalize/rdoc/String.html) related specs.

```
$ bundle exec mspec -t /Users/Juan/rubies/bin/ruby core/string/unicode_normalize_spec.rb
String#unicode_normalize
- returns nfc normalized form of string
- returns nfd normalized form of string
- returns nfkc normalized form of string
- returns nfkd normalized form of string
- raises an Encoding::CompatibilityError if string is not unicode encoding

String#unicode_normalize!
- modified original string

unicode_normalized?
- returns true if str is in Unicode normalization form
```

Ref. http://ruby-doc.org/stdlib-2.2.0/libdoc/unicode_normalize/rdoc/String.html

eregon commented:
Looks good!
Interesting, this is indeed in core (although ruby-doc.org says stdlib) as unicode_normalize is required at start-up in MRI prelude.

eregon commented on core/string/unicode_normalize_spec.rb:
```diff
@@ -0,0 +1,45 @@
+ruby_version_is "2.2" do
+  describe "String#unicode_normalize" do
+    it "returns nfc normalized form of string" do
+      "a\u0300".unicode_normalize.should == "à"
```
I think it's useful here to also have the assertion when explicitly passing :nfc (the default argument).

eregon commented:
"à" is the same in nfc/nkfc and in nfd/nkfd.
Maybe on of the characters mentioned at http://unicode.org/faq/normalization.html#6 could be used to make sure all normalizations behave as expected?

eregon commented on core/string/unicode_normalize_spec.rb:
```diff
@@ -0,0 +1,45 @@
+      "a\u0300".unicode_normalize(:nfkd).should == "à"
+    end
+
+    it "raises an Encoding::CompatibilityError if string is not unicode encoding" do
```
"is not *in an* unicode encoding"

eregon commented on core/string/unicode_normalize_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+    it "raises an Encoding::CompatibilityError if string is not unicode encoding" do
+      lambda do
+        eval("\xE0".force_encoding("ISO-8859-1").unicode_normalize(:nfd))
```
Useless eval I belive :wink:

eregon commented on core/string/unicode_normalize_spec.rb:
```diff
@@ -0,0 +1,45 @@
+
+      str.unicode_normalize!
+      str.unicode_normalized?.should be_true
+    end
```
This should have other examples/assertions with different normalizations (argument)

eregon commented:
The first line of a spec file must always require the `spec_helper`.
One way to ensure that and make sure the file organization is right is to use `mkspec`, mentioned in CONTRIBUTING.md.
In this particular case, move the file of this spec (at the root of rubyspec for instance) and then run:

     ../mspec/bin/mkspec -c String

and copy you changes in.

JuanitoFatas commented:
I am also seeing these errors on local, but they are the same:

```
1)
String#unicode_normalize returns normalized form of string by default 03D3 (ϓ) GREEK UPSILON WITH ACUTE AND HOOK SYMBOL FAILED
Expected "ϓ"
 to equal "ϓ"
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:17:in `block (4 levels) in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:4:in `block in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:3:in `<top (required)>'
                                                                                            
2)
String#unicode_normalize returns normalized form of string by default 03D4 (ϔ) GREEK UPSILON WITH DIAERESIS AND HOOK SYMBOL FAILED
Expected "ϔ"
 to equal "ϔ"
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:26:in `block (4 levels) in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:4:in `block in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:3:in `<top (required)>'
                                                                                            
3)
String#unicode_normalize returns normalized form of string by default 1E9B (ẛ) LATIN SMALL LETTER LONG S WITH DOT ABOVE FAILED
Expected "ẛ"
 to equal "ẛ"
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:35:in `block (4 levels) in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:4:in `block in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:3:in `<top (required)>'
                                                                                            
4)
String#unicode_normalize! modified original string (nfd) FAILED
Expected "aà"
 to equal "aà"
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:56:in `block (3 levels) in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:42:in `block in <top (required)>'
/home/travis/build/ruby/rubyspec/core/string/unicode_normalize_spec.rb:3:in `<top (required)>'
                                                                                            
5)
String#unicode_normalize! modified original string (nfkd) FAILED
Expected "aà"
 to equal "aà"
```

JuanitoFatas commented:
> The first line of a spec file must always require the spec_helper.
> One way to ensure that and make sure the file organization is right is to use mkspec, mentioned in CONTRIBUTING.md.
> In this particular case, move the file of this spec (at the root of rubyspec for instance) and then run:
> 
>  ../mspec/bin/mkspec -c String
> and copy you changes in.

Thanks for the heads up! Will do it next time.

eregon commented on core/string/unicode_normalized_spec.rb:
```diff
@@ -0,0 +1,105 @@
+require File.expand_path('../../../spec_helper', __FILE__)
```
There should be `# encoding: utf-8` since you use unicode string literals after (mostly useful for older versions which are not UTF-8 by default).

JuanitoFatas replied:
:+1: Added.

eregon commented on core/string/unicode_normalized_spec.rb:
```diff
@@ -0,0 +1,105 @@
+        unicode_03d3 = 'ϓ'
+
+        "\u03D3".unicode_normalize(:nfc).should        == unicode_03d3
+        "\u03D2\u0301".unicode_normalize(:nfd).should  == unicode_03d3
```
The assertion should be the other way around. And different nomalization forms are not `==` (or == would become *very* expensive).
```ruby
 "\u03D3".unicode_normalize(:nfd) == "\u03D2\u0301" # => true
```

eregon commented on core/string/unicode_normalized_spec.rb:
```diff
@@ -0,0 +1,105 @@
+    # http://unicode.org/faq/normalization.html#6
+    context "returns normalized form of string by default" do
+      it "03D3 (ϓ) GREEK UPSILON WITH ACUTE AND HOOK SYMBOL" do
+        unicode_03d3 = 'ϓ'
```
I think having the `"\u03D3"` notation would be clearer to see which in form this one is.
Probably the best is just to repeat `"\u03D3"` as the receiver in all examples for clarity.

eregon commented on core/string/unicode_normalized_spec.rb:
```diff
@@ -0,0 +1,105 @@
+  end
+
+  describe "String#unicode_normalize!" do
+    it "modified original string (nfc)" do
```
modifie*s*

eregon replied:
Other specs use `modifies self in place` which is good.

eregon commented:
Getting there :smiley:, please address my comments and make sure the CI passes.

eregon commented:
@JuanitoFatas Thanks a lot, that's a great addition!

JuanitoFatas commented:
@eregon Thank you for great code review :blush: !
